cleanrl/sac_continuous_action.py

```diff
@@ -192,88 +249,116 @@ def sample(self, n):
+        qf2_a_values = qf2.forward(s_obs, torch.Tensor(s_actions).to(device)).view(-1)
+        qf1_loss = loss_fn(qf1_a_values, next_q_value)
+        qf2_loss = loss_fn(qf2_a_values, next_q_value)
+
+        qf_loss = (qf1_loss + qf2_loss) / 2
+
+        values_optimizer.zero_grad()
+        qf_loss.backward()
+        values_optimizer.step()
+
+        if global_step % args.policy_frequency == 0: # TD 3 Delayed update support
+            for _ in range(args.policy_frequency): # compensate for the delay by doing 'actor_update_interval' instead of 1
+                pi, log_pi, _ = pg.get_action(s_obs)
+                qf1_pi = qf1.forward(s_obs, pi)
+                qf2_pi = qf2.forward(s_obs, pi)
+                min_qf_pi = torch.min(qf1_pi, qf2_pi).view(-1)
+
+                policy_loss = ((alpha * log_pi) - min_qf_pi).mean()
+
+                policy_optimizer.zero_grad()
+                policy_loss.backward()
+                policy_optimizer.step()
+
+                if args.autotune:
+                    with torch.no_grad():
+                        _, log_pi, _ = pg.get_action( s_obs)
+                    alpha_loss = ( -log_alpha * (log_pi + target_entropy)).mean()
+
+                    a_optimizer.zero_grad()
+                    alpha_loss.backward()
+                    a_optimizer.step()
+
+                    alpha = log_alpha.exp().item()
+
+        # update the target network
+        if global_step % args.target_network_frequency == 0:
             for param, target_param in zip(qf1.parameters(), qf1_target.parameters()):
                 target_param.data.copy_(args.tau * param.data + (1 - args.tau) * target_param.data)
             for param, target_param in zip(qf2.parameters(), qf2_target.parameters()):
                 target_param.data.copy_(args.tau * param.data + (1 - args.tau) * target_param.data)
-            qf1_losses[step] = qf1_loss.item()
-            qf2_losses[step] = qf2_loss.item()
-            policy_losses[step] = policy_loss.item()
-
-        if dones[step]:
-            break
-
-    # TRY NOT TO MODIFY: record rewards for plotting purposes
-    print(f"global_step={global_step}, episode_reward={rewards.sum()}")
-    writer.add_scalar("charts/episode_reward", rewards.sum(), global_step)
-    writer.add_scalar("losses/entropy", entropys[:step+1].mean().item(), global_step)
-    writer.add_scalar("losses/soft_q_value_1_loss", qf1_losses[:step+1].mean(), global_step)
-    writer.add_scalar("losses/soft_q_value_2_loss", qf2_losses[:step+1].mean(), global_step)
-    writer.add_scalar("losses/policy_loss", policy_losses[:step+1].mean(), global_step)
+
+    if done:
+        global_episode += 1 # Outside the loop already means the epsiode is done
+
+        writer.add_scalar("charts/episode_reward", episode_reward, global_step)
+        writer.add_scalar("charts/episode_length", episode_length, global_step)
+
+        if len(rb.buffer) > args.batch_size: # Makes sure loss and such are actually available.
```

Comment:
Consider moving the metrics recording scripts out of the`if done:` block, Additionally only record the loss every 100 times step. See example https://github.com/vwxyzjn/cleanrl/blob/65d0e4a1c3be60fd590b30108869b764990a2cb3/cleanrl/td3_continuous_action.py#L244